fdbserver/workloads/ReadHotDetection.actor.cpp

```diff
@@ -0,0 +1,158 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbrpc/ContinuousSample.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/workloads/BulkSetup.actor.h"
+#include "fdbclient/ReadYourWrites.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+struct ReadHotDetectionWorkload : TestWorkload {
+	int actorCount, keyCount;
+
+	double testDuration, transactionsPerSecond;
+	vector<Future<Void>> clients;
+	Key readKey;
+	KeyRange wholeRange;
+
+	ReadHotDetectionWorkload(WorkloadContext const& wcx) : TestWorkload(wcx) {
+		testDuration = getOption(options, LiteralStringRef("testDuration"), 120.0);
+		transactionsPerSecond = getOption(options, LiteralStringRef("transactionsPerSecond"), 1000.0) / clientCount;
+		actorCount = getOption(options, LiteralStringRef("actorsPerClient"), transactionsPerSecond / 5);
+		keyCount = getOption(options, LiteralStringRef("keyCount"), 100);
+		readKey = StringRef(format("testkey%08x", deterministicRandom()->randomInt(0, keyCount)));
+	}
+
+	virtual std::string description() { return "ReadHotDetection"; }
+
+	virtual Future<Void> setup(Database const& cx) { return _setup(cx, this); }
+
+	virtual Future<Void> start(Database const& cx) {
+		for (int c = 0; c < actorCount; c++) {
+			clients.push_back(timeout(keyReader(cx->clone(), this, actorCount / transactionsPerSecond,
+			                                    deterministicRandom()->random01() > 0.4),
+			                          testDuration, Void()));
+		}
+		return delay(testDuration);
+	}
+
+	virtual Future<bool> check(Database const& cx) {
+		if (clientId != 0) return true;
+		return _check(cx, this);
+	}
+
+	ACTOR Future<Void> _setup(Database cx, ReadHotDetectionWorkload* self) {
+		state int g = 0;
+		state Standalone<StringRef> largeValue;
+		state Standalone<StringRef> smallValue;
+		largeValue = self->randomString(largeValue.arena(), 100000);
+		smallValue = self->randomString(smallValue.arena(), 100);
+		state ReadYourWritesTransaction tr(cx);
+		loop {
+			try {
+				for (int i = 0; i < self->keyCount; i++) {
+					auto key = StringRef(format("testkey%08x", i));
+					if (key == self->readKey) {
+						tr.set(key, largeValue);
+					} else {
+						tr.set(key, deterministicRandom()->random01() > 0.8 ? largeValue : smallValue);
+					}
+				}
+				wait(tr.commit());
+				break;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+		self->wholeRange = KeyRangeRef(LiteralStringRef(""), LiteralStringRef("\xff"));
+		// TraceEvent("RHDLog").detail("Phase", "DoneSetup");
+		return Void();
+	}
+
+	ACTOR Future<bool> _check(Database cx, ReadHotDetectionWorkload* self) {
```

Comment:
check the read hot ranges during the run phase so that it is concurrent with the actual hot traffic

Reply:
Done.